ambry-api/src/main/java/com/github/ambry/compression/Compression.java

```diff
@@ -0,0 +1,220 @@
+   *
+   * Example of compatible upgrade:
+   *   Say today we use LZ4 1.8 and its name is "LZ4".  One year later, LZ4 1.9 is available.
+   *   Since LZ4 1.9 decompress() method supports decompressing LZ4 1.8 and 1.9, leave the name unchanged as "LZ4"
+   *   and just upgrade the LZ4 library to 1.9.
+   *
+   * Example of incompatible upgrade:
+   *   Say today we added LZ4 1.8 and its name is "LZ4".  One year later, LZ4 2.0 is released, but LZ4 2.0 is
+   *   incompatible with 1.8.  That means LZ4 2.0 cannot data compressed with L4 1.8.
+   *   In this case, introduce a new name like "LZ4_2" while keeping LZ4 1.8.
+   *
+   * @return A unique name of this compression algorithm.
+   */
+  String getAlgorithmName();
+
+  /**
+   * Calculate and return the size of buffer required to call compress().
+   * The caller is expected to allocate buffer of this size before calling compress().
+   *
+   * @param sourceDataSize The size of source data.
+   * @return The estimated buffer size required to store the compressed buffer in worst-case scenario.
+   */
+  int getCompressBufferSize(int sourceDataSize);
+
+  /**
+   * Get the original data size stored inside the shared/composite compressed buffer.  The compressed buffer
+   * contains version, algorithm name, original data size, and compressed data.
+   * The compressed buffer may be shared with other buffers, meaning only portion of this buffer was used by
+   * compression.  That is why it accepts the offset and size parameters.
+   * @param compressedBuffer The compressed buffer.
+   * @param offset The offset in the compressed buffer where compression data is stored.
+   * @param size The size in compressedBuffer to read.
+   * @return The size of original data.
+   */
+  int getDecompressBufferSize(byte[] compressedBuffer, int offset, int size);
+
+  /**
+   * Get the original data size stored in compressed buffer.  Compressed buffer is not shared and contains
+   * compressed data only.  The compressed buffer is the output of compress() that contains original/source data size.
+   * @param compressedBuffer The compressed buffer.
+   * @return Size of the original data.
+   */
+  default int getDecompressBufferSize(byte[] compressedBuffer) {
+    if (compressedBuffer == null) {
+      throw new IllegalArgumentException("compressedBuffer cannot be null.");
+    }
+    return getDecompressBufferSize(compressedBuffer, 0, compressedBuffer.length);
+  }
+
+  /**
+   * Compress portion of the shared buffer specified in {@code sourceData}.  Both {@code sourceData}
+   * and {@code compressedBuffer} can be shared/composite buffers.  That's why both buffers need to provide
+   * their size and offset.  The caller is responsible for managing memory allocation.
+   *
+   * The compressed buffer is a structure in this format:
+   * - Version (1 byte) - the structure/format version.  It starts with value '1'.
```

Comment:
Would we also need to encode the compression level along with the algorithm name?